src/Reflex/Dynamic.hs

```diff
@@ -243,7 +245,21 @@ joinDynThroughMap = joinDyn . fmap distributeMapOverDynPure
 -- | Combine a 'Dynamic' of a 'Traversable' of 'Dynamic's into a 'Dynamic'
 -- with the current values of the 'Dynamic's in the functor.
 joinDynThroughTraversable :: (Reflex t, Traversable f) => Dynamic t (f (Dynamic t a)) -> Dynamic t (f a)
-joinDynThroughTraversable = join . fmap sequence
+joinDynThroughTraversable = join . fmap (unsafeViaList distributeListOverDynPure)
+  where
+    -- Requires 'f' to preserve number and order of elements
+    unsafeViaList :: (Functor f, Traversable t) => ([a] -> f [b]) -> t a -> f (t b)
+    unsafeViaList f t = unsafeReifyContents t <$> f (toList t)
+
+    -- Requires the list to have at least as many elements as the traversable
+    unsafeReifyContents :: Traversable t => t a -> [b] -> t b
+    unsafeReifyContents = evalState . sequence . (f <$)
+      where
+       f = do
+         (b:bs') <- get
+         put bs'
+         return b
+
```

Comment:
Let's make the failure cases explicit, with `error` calls: 1) if `get` returns an empty list, or 2) if anything is left over after the state monad is done running.

Reply:
Would it be better to instead error out at the start if the traversable and list do not have the same length?

Reply:
@Ericson2314 , @ryantrinkle any thoughts on the question above?